test/test_tagging.py

```diff
@@ -0,0 +1,241 @@
+import json
+from typing import (
+    Optional,
+)
+import unittest
+from unittest.mock import (
+    MagicMock,
+    patch,
+)
+
+from azul.deployment import (
+    populate_tags,
+)
+from azul.logging import (
+    configure_test_logging,
+)
+from azul_test_case import (
+    AzulUnitTestCase,
+)
+
+
+def setupModule():
+    configure_test_logging()
+
+
+class TestTerraformResourceTags(AzulUnitTestCase):
+
+    def permissive_compare(self, expected, actual):
+        """
+        Wrapper for assertDictEqualPermissive doctests.
+
+        >>> t = TestTerraformResourceTags()
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 789}]},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 789}]}
+        ... )
+
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: None}]},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'abc'}]}
+        ... )
+
+        >>> t.permissive_compare(  #doctest: +ELLIPSIS
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'def'}]},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'abc'}]}
+        ... )
+        ...
+        Traceback (most recent call last):
+        ...
+        AssertionError: ... Discrepancy in path: .qaz[0].456
+
+        >>> t.permissive_compare(  #doctest: +ELLIPSIS
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'def'}]},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: None}]}
+        ... )
+        Traceback (most recent call last):
+        ...
+        AssertionError: ... Discrepancy in path: .qaz[0].456
+
+        >>> t.permissive_compare(  #doctest: +ELLIPSIS
+        ...     {'foo': 'bar', 'qaz': None, 456: 'def'},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'abc'}]}
+        ... )
+        Traceback (most recent call last):
+        ...
+        AssertionError: ... Discrepancy in path: object length mismatch
+
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': None},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'abc'}]}
+        ... )
+        """
+        path = self.assertDictEqualPermissive(expected, actual)
+        return self.assertIsNone(path, f'Discrepancy in path: {path}')
+
+    def assertDictEqualPermissive(self, expected, actual, path: str = '') -> Optional[str]:
```

Comment:
```suggestion
    def assertDictEqualPermissive(self, expected, actual, path: Tuple[str] = ()) -> Optional[str]:
```

or

```suggestion
    def assertDictEqualPermissive(self, expected, actual, *path: str) -> Optional[str]:
```

Whichever is more convenient.

Reply:
This method compares. It does not assert. Hence it should not have "assert" in its name.